### PR TITLE
Optimize size(dim) and stride(dim)

### PR DESCRIPTION
**Pull request by peterbell10**
This improves `c10::maybe_wrap_dim` to short-cut the "happy path"
where dim is in the correct range, and also moves the error and scalar
edge-cases out-of-line. These changes cut callgrind instruction counts
for `size(i)` from 5200 to 2000.

In the `size` and `stride` methods themselves, I also avoid calling
`TensorImpl::dim()` since it may be a virtual call. This further
reduced the instruction count from 2000 to 1500.

For comparison, `tensor.sizes()[0]` takes 1200 instructions so
`tensor.size(0)` is still marginally slower. This is unavoidable
though since it has to handle dimension wrapping.

### CODE REVIEW

File: c10/core/WrapDimMinimal.h [hunk 1]
```diff
@@ -4,37 +4,22 @@
...
-        ")");
+  // Inline the fast paths
+  if (C10_LIKELY(-dim_post_expr <= dim && dim < dim_post_expr)) {
+    // Branch-less version of dim + (dim < 0 ? dim_post_expr : 0)
+    return dim + dim_post_expr * (dim < 0);
   }
-  if (dim < 0)
-    dim += dim_post_expr;
-  return dim;
+  // Check for edge-cases out-of-line
```
Comment (Lezcano) [inline c10/core/WrapDimMinimal.h:21]: Please add that the edge cases are `dim_post_expr <= 0` and the `wrap_scalar` argument to the comment.

File: c10/core/WrapDimMinimal.cpp [hunk 2]
```diff
@@ -0,0 +1,36 @@
+#include <c10/core/WrapDimMinimal.h>
+
+namespace c10 {
+namespace detail {
+
+int64_t maybe_wrap_dim_slow(
+    int64_t dim,
+    int64_t dim_post_expr,
+    bool wrap_scalar) {
+  if (dim_post_expr <= 0) {
```
Comment (Lezcano) [inline c10/core/WrapDimMinimal.cpp:10]: Very odd that this is `<= 0`. @ezyang is this intended? I would assume that `dim_post_expr` must be nonnegative.

File: c10/core/WrapDimMinimal.cpp [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+      "Dimension out of range (expected to be in range of [",
+      min,
+      ", ",
+      max,
+      "], but got ",
+      dim,
+      ")");
+  if (dim < 0)
+    dim += dim_post_expr;
+  return dim;
```
Comment (Lezcano) [inline c10/core/WrapDimMinimal.cpp:32]: This should be an unconditional error now, right?

**Review by lezcano — COMMENTED** [review]
Good stuff.

File: c10/core/WrapDimMinimal.cpp (hunk 3 above)
Comment (peterbell10) [reply c10/core/WrapDimMinimal.cpp]: Changed to `TORCH_INTERNAL_ASSERT(false)`.

Comment (Lezcano) [reply c10/core/WrapDimMinimal.cpp]: I think that asan will complain about htis function not having a return, but we'll see.

**Review by lezcano — APPROVED** [review]
LGTM!

File: c10/core/WrapDimMinimal.h [hunk 4]
```diff
@@ -4,37 +4,22 @@
...
-        false,
-        "Dimension out of range (expected to be in range of [",
-        min,
-        ", ",
-        max,
-        "], but got ",
-        dim,
-        ")");
+  // Inline the fast paths
+  if (C10_LIKELY(-dim_post_expr <= dim && dim < dim_post_expr)) {
```
Comment (peterbell10) [inline c10/core/WrapDimMinimal.h:17]: I've checked and an `unsafe_size(dim)` could bring the cost down to `1200` instructions by removing the bounds checking. However, I'm erring on the side of "if you care this much about overhead you should be calling `.sizes()`directly so you can re-use the same ArrayRef".

**Review by ngimel — APPROVED** [review]

File: c10/core/WrapDimMinimal.h (hunk 4 above)
Comment (Lezcano) [reply c10/core/WrapDimMinimal.h:17]: I still like the `unsafe_size(dim)` idea, as it's something that comes up a lot. @ezyang wdyt?

File: c10/core/WrapDimMinimal.h [hunk 5]
```diff
@@ -4,37 +4,22 @@
...
+    // Branch-less version of dim + (dim < 0 ? dim_post_expr : 0)
```
Comment (swolchok) [inline c10/core/WrapDimMinimal.h:18]: the two versions of clang I tried generate the same assembly (using a branch) either way -- https://godbolt.org/z/xPG8M1nWr

Comment (peterbell10) [reply c10/core/WrapDimMinimal.h:18]: The branching in that assembly is for the `if` statement above (which is likely to be predicted correctly, so okay to branch). 

Clang is making a similar optimization as me to make the ternary version branch-less
```
        mov     rax, rdi
        sar     rax, 63 
        and     rax, rsi
        add     rax, rdi
```

This extends the sign bit of `dim` to fill the entirety of `rax`, then does `dim + (dim_post_expr & rax)`.

**Pull request closed without merging**